src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -108,219 +287,397 @@ public virtual async Task InvokeAsync(HttpContext context, RequestDelegate next)
+                }
+                await Task.WhenAll(resultTasks);
+                for (int i = 0; i < gqlRequests.Count; i++)
+                {
+                    results[i] = await resultTasks[i];
+                }
+            }
+            else
+            {
+                for (int i = 0; i < gqlRequests.Count; i++)
+                {
+                    results[i] = await ExecuteRequestAsync(context, gqlRequests[i], context.RequestServices, userContext);
+                }
             }
-
-            await WriteResponseAsync(context.Response, _serializer, cancellationToken, executionResults);
         }
+        await WriteJsonResponseAsync(context, HttpStatusCode.OK, results);
     }
 
-    protected virtual async ValueTask<bool> HandleDeserializationErrorAsync(HttpContext context, Exception ex)
+    /// <summary>
+    /// Executes a GraphQL request with a scoped service provider.
+    /// <br/><br/>
+    /// Typically this method should create a service scope and call
+    /// <see cref="ExecuteRequestAsync(HttpContext, GraphQLRequest, IServiceProvider, IDictionary{string, object?})">ExecuteRequestAsync</see>,
+    /// disposing of the scope when the asynchronous operation completes.
+    /// </summary>
+    protected abstract Task<ExecutionResult> ExecuteScopedRequestAsync(HttpContext context, GraphQLRequest? request, IDictionary<string, object?> userContext);
+
+    /// <summary>
+    /// Executes a GraphQL request.
+    /// <br/><br/>
+    /// It is suggested to use the <see cref="HttpGetValidationRule"/> and
+    /// <see cref="HttpPostValidationRule"/> to ensure that only query operations
+    /// are executed for GET requests, and query or mutation operations for
+    /// POST requests.
+    /// This should be set in both <see cref="ExecutionOptions.ValidationRules"/> and
+    /// <see cref="ExecutionOptions.CachedDocumentValidationRules"/>, as shown below:
+    /// <code>
+    /// var rule = isGet ? new HttpGetValidationRule() : new HttpPostValidationRule();
+    /// options.ValidationRules = DocumentValidator.CoreRules.Append(rule);
+    /// options.CachedDocumentValidationRules = new[] { rule };
+    /// </code>
+    /// </summary>
+    protected abstract Task<ExecutionResult> ExecuteRequestAsync(HttpContext context, GraphQLRequest? request, IServiceProvider serviceProvider, IDictionary<string, object?> userContext);
+
+    /// <summary>
+    /// Builds the user context based on a <see cref="HttpContext"/>.
+    /// <br/><br/>
+    /// Note that for batch or WebSocket requests, the user context is created once
+    /// and re-used for each GraphQL request or data event that applies to the same
+    /// <see cref="HttpContext"/>.
+    /// <br/><br/>
+    /// To tailor the user context individually for each request, call
+    /// <see cref="global::GraphQL.GraphQLBuilderExtensions.ConfigureExecutionOptions(IGraphQLBuilder, Action{ExecutionOptions})"/>
```

Comment:
```suggestion
    /// <see cref="GraphQLBuilderExtensions.ConfigureExecutionOptions(DI.IGraphQLBuilder, Action{ExecutionOptions})"/>
```

Reply:
Interesting.  Reference must be different here than in my library for some reason.